### DISCUSSION TITLE
Is there a wiki for adding new Hardware/Sensor/Driver? How to publish Mqtt Messages?

### DISCUSSION THREAD

**Discussion by Failx92** [arendst/Tasmota#19192] [Q&A]
Hello,

is there a wiki on how to create new sensors/hardware? 

I am currently trying to integrate an [SD-WIFI adapter](https://github.com/FYSETC/FYSETC-SD-WIFI) into Tastmota and it is really difficult for me to find a start.

It took me hours to figure out how to add a custom GPIO Pin and call my code. 

At the moment I have problems how to transfer data to mqtt/tele.

How to publish Mqtt Messages? 

There are some functions (Response_P/ResponseAppend_P) but I don't understand how to use them, the syntax confuses me.  Can someone give me a hint? 

Greetings

**Comment by barbudor** [comment]
Have you looked at https://tasmota.github.io/docs/For-Developers/

Beside that, there are a lot of existing drivers that can be used as samples

Publishing MQTT messages may depends what you want to do
For exemple, adding keys to the standard teleperiod SENSOR message is done from a callback, while publishing ad-hoc messages a be done directly

Look for some drivers that may do similar things to what you want to do

Now, for the FYSETC you don't need to write any code
SD-CARD file systzis already supported and you can upload/download files through the web UI
The only thing is that you should take care for not accessing the card through tasmota at the same time as the main device as the hardware do not provide any security against conflicts

I think @Jason2866 has played a bit with that and can share his configuration and' experience
I have one I a drawer. Still haven't had time to pay with it

**Reply by Jason2866** [reply]
What's your issue, at all? Does the SD Card not work? What do you want to achieve?
Generating mqtt message is not device related, it is a general question.

**Reply by Failx92** [reply]
Yes I looked at the docs but haven't found something usefull for my problem,  so I thought it was written elsewhere. 

I have tried to copy the Mqtt call for an existing sensor, but I seem to have syntax problems. Because when the ESP is supposed to send Data via "FUNC_JSON_APPEND", it crashes. So I have to debug but I don't understand the syntax when which "/"" whatever is expected. 

For the FYSETC I have to write code because I need to take care for the device in which it is insert. Thats not the problem this part of the code works without conflicts.

Background: I've got a Dostmann Air Control 5000 which logs all data to the sd card. Then I read this data with the esp and want to send it to iobroker via mqtt.

**Reply by barbudor** [reply]
Do you have a fork/branch that is visible in Github so we can have a look at your code and advise?

Processing FUNC_JSON_APPEND callback should be very straight forward as you simply have to append json keys but actually no check is performed on the json syntax
However, your data may be too long?

I suggest some basic debugging steps
- I serting a simple hardcoded fixed key/value pair to check you have the mecanism right
- adding step by step real values


Note that the crash may be related to a too long payload that exhaust memory allocation.
Do you have an estimate of the expected resulting JSON size?
